File: android-core/src/main/java/com/mparticle/internal/messages/BaseMPMessageBuilder.java
package com.mparticle.internal.messages;


import android.location.Location;
import androidx.annotation.Nullable;

import com.mparticle.internal.Constants;
import com.mparticle.internal.InternalSession;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class BaseMPMessageBuilder extends JSONObject {
    private Double mLength = null;

    public BaseMPMessageBuilder(String messageType) {
        try {
            put(Constants.MessageKey.TYPE, messageType);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public BaseMPMessageBuilder timestamp(long timestamp) {
        try {
            put(Constants.MessageKey.TIMESTAMP, timestamp);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return this;
    }

    public Long getTimestamp() {
        long timestamp = optLong(Constants.MessageKey.TIMESTAMP, -1);
        if (timestamp != -1) {
            return timestamp;
        } else {
            return null;
        }
    }

    public BaseMPMessageBuilder name(String name) {
        try {
            put(Constants.MessageKey.NAME, name);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return this;
    }

    public BaseMPMessageBuilder attributes(JSONObject attributes) {
        if (attributes != null && attributes.length() > 0) {
            try {
                put(Constants.MessageKey.ATTRIBUTES, attributes);
                if (mLength != null) {
                    addEventLengthAttributes(mLength);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return this;
    }

    public String getMessageType() {
        return optString(Constants.MessageKey.TYPE);
    }

    public BaseMPMessage build(InternalSession session, @Nullable Location location, long mpId) throws JSONException {
        return new BaseMPMessage(this, session, location, mpId);
    }

    public BaseMPMessageBuilder dataConnection(String dataConnection) {
        if (dataConnection != null) {
            try {
                put(Constants.MessageKey.STATE_INFO_DATA_CONNECTION, dataConnection);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return this;
    }

    public BaseMPMessageBuilder length(Double length) {
        mLength = length;
        if (length != null){
            try {
                put(Constants.MessageKey.EVENT_DURATION, length);
                addEventLengthAttributes(length);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return this;
    }

    private void addEventLengthAttributes(Double length) {
        try {
            if (!has(Constants.MessageKey.ATTRIBUTES)) {
                put(Constants.MessageKey.ATTRIBUTES, new JSONObject());
            }
            if (!getJSONObject(Constants.MessageKey.ATTRIBUTES).has("EventLength")) {
                //can't be longer than max int milliseconds
                getJSONObject(Constants.MessageKey.ATTRIBUTES).put("EventLength", Integer.toString(length.intValue()));
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public BaseMPMessageBuilder flags(Map<String, List<String>> customFlags) {
        if (customFlags != null) {
            try {

                JSONObject flagsObject = new JSONObject();
                for (Map.Entry<String, List<String>> entry : customFlags.entrySet()) {
                    List<String> values = entry.getValue();
                    JSONArray valueArray = new JSONArray(values);
                    flagsObject.put(entry.getKey(), valueArray);
                }
                put(Constants.MessageKey.EVENT_FLAGS, flagsObject);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return this;
    }

    String[] getKeys() {
        List<String> strings = new ArrayList();
        Iterator<String> iterator = keys();
        while (iterator.hasNext()) {
            strings.add(iterator.next());
        }
        return strings.toArray(new String[0]);
    }
}

File: android-core/src/main/kotlin/com/mparticle/internal/messages/BaseMPMessageBuilder.kt
package com.mparticle.internal.messages

import android.location.Location
import com.mparticle.internal.Constants
import com.mparticle.internal.InternalSession
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject


open class BaseMPMessageBuilder(messageType: String) : JSONObject() {
    private var mLength: Double? = null

    init {
        try {
            put(Constants.MessageKey.TYPE, messageType)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }


    open fun timestamp(timestamp: Long): BaseMPMessageBuilder {
        try {
            put(Constants.MessageKey.TIMESTAMP, timestamp)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return this
    }

    val timestamp: Long?
        get() {
            val timestamp = optLong(Constants.MessageKey.TIMESTAMP, -1)
            return if (timestamp != -1L) {
                timestamp
            } else {
                null
            }
        }

    fun name(name: String): BaseMPMessageBuilder {
        try {
            put(Constants.MessageKey.NAME, name)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return this
    }

    fun attributes(attributes: JSONObject?): BaseMPMessageBuilder {
        if (attributes != null && attributes.length() > 0) {
            try {
                put(Constants.MessageKey.ATTRIBUTES, attributes)
                mLength?.let { addEventLengthAttributes(it) }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return this
    }

    val messageType: String
        get() = optString(Constants.MessageKey.TYPE)

    @Throws(JSONException::class)
    open fun build(session: InternalSession, location: Location?, mpId: Long): BaseMPMessage {
        return BaseMPMessage(this, session, location, mpId)
    }

    fun dataConnection(dataConnection: String): BaseMPMessageBuilder {
        try {
            put(Constants.MessageKey.STATE_INFO_DATA_CONNECTION, dataConnection)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return this
    }

    fun length(length: Double?): BaseMPMessageBuilder {
        mLength = length
        if (length != null) {
            try {
                put(Constants.MessageKey.EVENT_DURATION, length)
                addEventLengthAttributes(length)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return this
    }

    private fun addEventLengthAttributes(length: Double) {
        try {
            if (!has(Constants.MessageKey.ATTRIBUTES)) {
                put(Constants.MessageKey.ATTRIBUTES, JSONObject())
            }
            if (!getJSONObject(Constants.MessageKey.ATTRIBUTES).has("EventLength")) {
                //can't be longer than max int milliseconds
                getJSONObject(Constants.MessageKey.ATTRIBUTES).put("EventLength", length.toInt().toString())
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    fun flags(customFlags: Map<String?, List<String?>>?): BaseMPMessageBuilder {
        if (customFlags != null) {
            try {
                val flagsObject = JSONObject()
                for ((key, values) in customFlags) {
                    val valueArray = JSONArray(values)
                    flagsObject.put(key, valueArray)
                }
                put(Constants.MessageKey.EVENT_FLAGS, flagsObject)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return this
    }

    val keys: Array<String?>
        get() {
            val strings: MutableList<String?> = ArrayList<String?>()
            val iterator = keys()
            while (iterator.hasNext()) {
                strings.add(iterator.next())
            }
            return strings.toTypedArray<String?>()
        }
}